feat: add comprehensive e2e tests for ApplicationSet progressive sync

AvhiMaz opened the pull request:
## Description
  
  This PR adds comprehensive end-to-end tests for the ApplicationSet progressive sync feature, addressing issue #25042.

  The implementation includes:
  - 5 e2e test functions covering various progressive sync scenarios
  - 2 new helper functions for progressive sync validation
  - Feature enablement via ConfigMap setting
  - Tests for 2-step, 3-step, and percentage-based rollout strategies
  - Tests for deletion strategies (AllAtOnce and Reverse order)

  ## Related Issues

 Fixes #25042

AvhiMaz committed:
feat: add comprehensive e2e tests for ApplicationSet progressive sync

  - Implement 5 e2e test functions covering progressive sync scenarios
  - Add ApplicationSetHasProgressiveStatus() for wave completion validation
  - Add OnlyApplicationsExist() for sequential order verification
  - Enable progressive sync feature via ConfigMap in test init()
  - Test 2-step, 3-step, and percentage-based maxUpdate constraints
  - Test deletion strategies (AllAtOnce and Reverse order)
  - Use Ginkgo BDD pattern with Given/When/Then fixtures

  Fixes #25042

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

AvhiMaz committed:
fix: resolve linting errors in progressive sync e2e tests

  - Handle error return value from SetParamInSettingConfigMap (errcheck)
  - Fix import ordering for gofumpt compliance

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

AvhiMaz committed:
Merge branch 'master' into feat/add-progressive-sync-e2e-tests-clean

AvhiMaz committed:
fix: remove empty And() blocks in progressive sync e2e tests

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

AvhiMaz committed:
fix: remove explicit namespace from ApplicationSet creation in progressive sync e2e tests

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

AvhiMaz committed:
fix: implement progressive sync filtering and correct test namespaces

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

AvhiMaz committed:
fix: fix lint error

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

nitishfy commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -232,6 +232,8 @@ func (r *ApplicationSetReconciler) Reconcile(ctx context.Context, req ctrl.Reque
 	// appSyncMap tracks which apps will be synced during this reconciliation.
 	appSyncMap := map[string]bool{}
+	var appDependencyList [][]string
+	var appStepMap map[string]int
```
better to put the `appSyncMap` too.

nitishfy commented on applicationset/controllers/applicationset_controller.go:
```diff
@@ -1590,6 +1594,65 @@ func (r *ApplicationSetReconciler) syncDesiredApplications(logCtx *log.Entry, ap
 
+// filterApplicationsByProgressiveStep filters applications to only include those in the current eligible step.
+// Applications are only created/updated if all applications in their current step's prerequisite steps are healthy.
+func (r *ApplicationSetReconciler) filterApplicationsByProgressiveStep(logCtx *log.Entry, appset *argov1alpha1.ApplicationSet, applications []argov1alpha1.Application, appDependencyList [][]string, appStepMap map[string]int) []argov1alpha1.Application {
```
please write unit tests for this function.

nitishfy reviewed (changes requested):
Thank you for raising the PR. I don't think the e2e tests have been written so far, do you intend to put this in draft state? Secondly, i would like to hear from @ranakan19 thoughts on this PR especially on the implementation of `filterApplicationsByProgressiveStep()` and if we actually need it. If everything goes well, please consider writing unit tests too. Thanks!

AvhiMaz committed:
fix: implement PR review feedback for progressive sync tests
  - Consolidate variable declarations
  - Add method documentation
  - Fix error return values
  - Add unit tests for filterApplicationsByProgressiveStep

Signed-off-by: AvhiMaz <avhimazumder5@outlook.com>

ranakan19 reviewed (changes requested):
Adding `filterApplicationsByProgressiveStep` is changing the functionality of how progressive sync behaves. Currently all the applications are created at once but it is the sync which is progressive. But with this change, it creates applications incrementally.
IMO that should be a separate PR and the scope of this PR should only be adding e2e tests for the existing functionality. 

I also faced timeout [errors](https://gist.github.com/ranakan19/7a0d5d997d1b279b85ac3af3349188d5) when running them locally. If you were able to run them locally did you have to change the timeout?
e2e tests are failing on this PR as well

ranakan19 commented:
I have a draft PR https://github.com/argoproj/argo-cd/pull/26092 for adding e2e tests - I've also included running it in ci in my PR. We run e2e tests as a precheck on new PRs, so it made sense to include it in e2e. But since progressive sync needs the feature to enabled at controller level, some work had to go in there.

ranakan19 commented:
closing this in favor of https://github.com/argoproj/argo-cd/pull/26092 and https://github.com/argoproj/argo-cd/pull/26673 - which have already been merged
